virtualizarr/writers/icechunk.py

```diff
@@ -96,65 +112,147 @@ def write_variables_to_icechunk_group(
-        chunk_shape=zarray.chunks,
-        dtype=encode_dtype(zarray.dtype),
-        codecs=zarray._v3_codec_pipeline(),
-        dimension_names=var.dims,
-        fill_value=zarray.fill_value,
-        # TODO fill_value?
-    )
+    dims: list[str] = cast(list[str], list(var.dims))
+    append_axis, existing_num_chunks, arr = None, None, None
+    if append_dim and append_dim not in dims:
+        raise ValueError(
+            f"append_dim {append_dim} not found in variable dimensions {dims}"
+        )
+    if mode == "a":
+        existing_array = group[name]
+        if not isinstance(existing_array, Array):
+            raise ValueError("Expected existing array to be a zarr.core.Array")
+        append_axis = get_axis(dims, append_dim)
+
+        # check if arrays can be concatenated
+        _check_compatible_arrays(ma, existing_array, append_axis)
+        _check_compatible_encodings(var.encoding, existing_array.attrs)
+
+        # determine number of existing chunks along the append axis
+        existing_num_chunks = num_chunks(
+            array=group[name],
+            axis=append_axis,
+        )
 
-    # TODO it would be nice if we could assign directly to the .attrs property
-    for k, v in var.attrs.items():
-        arr.attrs[k] = encode_zarr_attr_value(v)
+        # resize the array
+        arr = resize_array(
+            group=group,
+            name=name,
+            var=var,
+            append_axis=append_axis,
+        )
+    else:
+        # create array if it doesn't already exist
+        arr = group.require_array(
+            name=name,
+            shape=zarray.shape,
+            chunk_shape=zarray.chunks,
+            dtype=encode_dtype(zarray.dtype),
+            codecs=zarray._v3_codec_pipeline(),
+            dimension_names=var.dims,
+            fill_value=zarray.fill_value,
+        )
 
-    _encoding_keys = {"_FillValue", "missing_value", "scale_factor", "add_offset"}
-    for k, v in var.encoding.items():
-        if k in _encoding_keys:
+        # TODO it would be nice if we could assign directly to the .attrs property
+        for k, v in var.attrs.items():
             arr.attrs[k] = encode_zarr_attr_value(v)
 
+        _encoding_keys = {"_FillValue", "missing_value", "scale_factor", "add_offset"}
+        for k, v in var.encoding.items():
+            if k in _encoding_keys:
+                arr.attrs[k] = encode_zarr_attr_value(v)
+
     write_manifest_virtual_refs(
         store=store,
         group=group,
         arr_name=name,
         manifest=ma.manifest,
+        append_axis=append_axis,
+        existing_num_chunks=existing_num_chunks,
+    )
+
+
+def generate_chunk_key(
+    index: tuple[int, ...],
+    append_axis: Optional[int] = None,
+    existing_num_chunks: Optional[int] = None,
+) -> str:
+    if append_axis and append_axis >= len(index):
+        raise ValueError(
+            f"append_axis {append_axis} is greater than the number of indices {len(index)}"
+        )
```

Comment:
There's perhaps another potential bug if someone passes a negative number for the axis...

Reply:
The `dataset_to_icechunk` function accepts an append_dim by name - so I don't think the append_axis could be negative. It could not exist though which is why we have:

```
    if append_dim and append_dim not in dims:
        raise ValueError(
            f"append_dim {append_dim} not found in variable dimensions {dims}"
        )
```